core/src/main/kotlin/org/eln2/data/MultiMap.kt

```diff
@@ -0,0 +1,197 @@
+package org.eln2.data
+
+/**
+ * A helper class for representing a mapping as a Map.Entry.
+ */
+data class MultiMapEntry<K, V>(override val key: K, override val value: V): Map.Entry<K, V>
+
+/**
+ * A MultiMap: a data structure that maps one key (K) to zero or more values (V).
+ *
+ * This is in contrast to the usual Map, which maps one key to at most one value.
+ *
+ */
+interface MultiMap<K, V> {
+	/**
+	 * Tests whether a key [k] is in the MultiMap.
+	 *
+	 * This is true if k is associated with at least one value.
+	 */
+	operator fun contains(k: K): Boolean = get(k).isNotEmpty()
+
+	/**
+	 * Get the set of all values associated with key [k].
+	 */
+	operator fun get(k: K): Set<V>
+
+	/**
+	 * An iterator over every Entry<K, Set<V>>, of size keySize.
+	 */
+	val keyMapping: Iterable<Map.Entry<K, Set<V>>>
+	/**
+	 * The number of keys, value sets, and keyMapping entries in this MultiMap.
+	 */
+	val keyMappingSize: Int get() = keyMapping.map { 1 }.sum()
+	/**
+	 * The set of all keys.
+	 */
+	val keys: Set<K> get() = keyMapping.map {(k, _) -> k }.toSet()
+	/**
+	 * The set of all sets of values.
+	 * 
+	 * This will have the same size as keySize.
+	 */
+	val valueSets: Collection<Set<V>> get() = keyMapping.map {(_, vs) -> vs}
+	/**
+	 * The unique values in this MultiMap.
+	 */
+	val uniqueValues: Set<V>
+	/**
+	 * The number of unique values in this MultiMap.
+	 */
+	val uniqueValuesSize: Int get() = uniqueValues.size
+	/**
+	 * An iterator over every Entry<K, V>, where K may be repeated for each value.
+	 */
+	val entries: Iterable<Map.Entry<K, V>>
+	/**
+	 * The number of pairings in this MultiMap.
+	 */
```

Comment:
Probably want to add `O(n)` to the docs.

In fact, ideally add algorithmic complexity to every method. It's not always obvious; without reading the implementation, I'd assume a getter for a single int is always O(1).

Reply:
You do have a point... the current implementations override this to be O(1) anyway. This is just the _ultimate_ fallback--if you can do better than this, you should.

Reply:
If  _every_ implementation does better, then there's no need for the base case; it becomes nothing more than a potential (performance) bug in the future.

Reply:
I'm still of the mind of reducing implementation burden for the future, but in this case:
1. The implementation here is likely to be a footgun for a lazy implementor, and
2. This is the core codebase where we don't expect a lot of naive developers to be poking around.

I'll leave the current implementation in a comment as a suggestion, but you're right in that it might not be best as a _fallback_.